### PR TITLE
REF: Convert `Self` usages in "Convert struct to tuple" refactoring

### LINKED ISSUE #5851: Converting struct to tuple doesn't convert struct creation via Self
## Environment

* **IntelliJ Rust plugin version:** 0.3.127.3249-202
* **Rust toolchain version:** 1.4.5.0
* **IDE name and version:** CLion 2020.2
* **Operating system:** Linux

## Problem description
The "convert to tuple" intention doesn't refactor construction.

## Steps to reproduce
```rust
struct MyStruct {
    a: i32,
    b: i32,
}
impl MyStruct {
    fn new(a: i32, b: i32) -> Self {
        Self {a,b}
    }

    fn destructure(self) -> (i32, i32) {
        let MyStruct{a,b} = self;
        (a,b)
    }
}
```
Running "convert to tuple" intention gives the following:

```rust

struct MyStruct(i32, i32);

impl MyStruct {
    fn new(a: i32, b: i32) -> Self {
        Self {a,b} // <------ not converted! it should become `Self(a, b)`
    }

    fn destructure(self) -> (i32, i32) {
        let MyStruct(a, b) = self;
        (a,b)
    }
}
```

### PR DESCRIPTION
**Pull request by dima74**
Fixes #5851

changelog: Convert `Self` usages in `Convert struct to tuple` refactoring

### CODE REVIEW

**Review by ortem — APPROVED** [review]
Generally LGTM, but maybe it is also worth asking @vlad20012, particularly about the new `searchReferencesWithSelf` method

**Review by vlad20012 — APPROVED** [review]
Well, it's cool that it fixes the bug, but really I think this is a partial solution and it does not fixes the problem in general. I think any reference search query must search for `Self` usages automatically. This is possible using some extension points, but I don't know which exactly

**Comment by dima74** [comment]
> I think any reference search query must search for `Self` usages automatically. This is possible using some extension points, but I don't know which exactly

There is `RsReferencesSearchExtensionImpl`, but I can't find a way to reuse already found references with `QueryExecutorBase` api (so that we don't run base search references query twice)

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsElement.kt [hunk 1]
```diff
@@ -216,3 +216,39 @@ fun RsElement.searchReferences(scope: SearchScope? = null): Query<PsiReference>
...
+ *                 ~~~~ result
+ * }
+ */
+private fun PsiReference.findSelfReferences(): List<PsiReference>? {
+    val implPath = element as? RsPath ?: return null
+    val implType = implPath.parent as? RsPathType ?: return null
+    val impl = implType.parent as? RsImplItem ?: return null
+    if (impl.typeReference != implType) return null
+    return impl
+        .stubDescendantsOfTypeStrict<RsPath>()
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/psi/ext/RsElement.kt]: Fixed

**Review by vlad20012 — APPROVED** [review]
bors r+

**Pull request merged**